security/advancedtls/pemfile_provider_test.go

```diff
@@ -0,0 +1,220 @@
+				KeyFile:   test.keyFile,
+				CertFile:  test.certFile,
+				TrustFile: test.trustFile,
+			}
+			provider, err := NewPEMFileProvider(options)
+			if (err != nil) != test.expectedError {
+				t.Fatalf("got and want mismatch, want expectedError = %v, got %v", test.expectedError, err != nil)
+			}
+			if err != nil {
+				return
+			}
+			provider.Close()
+		})
+	}
+
+}
+
+// This test overwrites the credential reading function used by the watching goroutine. It is tested under different stages:
+// At stage 0, we force reading function to load clientPeer1 and serverTrust1, and see if the credentials are picked up by the watching go routine.
+// At stage 1, we force reading function to cause an error. The watching go routine should log the error while leaving the credentials unchanged.
+// At stage 2, we force reading function to load clientPeer2 and serverTrust2, and see if the new credentials are picked up.
+func (s) TestWatchingRoutineUpdates(t *testing.T) {
+	// Load certificates.
+	cs := &certStore{}
+	err := cs.loadCerts()
+	if err != nil {
+		t.Fatalf("cs.loadCerts() failed: %v", err)
+	}
+	tests := []struct {
+		desc         string
+		options      *PEMFileProviderOptions
+		wantKmStage0 certprovider.KeyMaterial
+		wantKmStage1 certprovider.KeyMaterial
+		wantKmStage2 certprovider.KeyMaterial
+	}{
+		{
+			desc: "use identity certs and root certs",
+			options: &PEMFileProviderOptions{
+				CertFile:  "not_empty_cert_file",
+				KeyFile:   "not_empty_key_file",
+				TrustFile: "not_empty_trust_file",
+			},
+			wantKmStage0: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}, Roots: cs.serverTrust1},
+			wantKmStage1: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}, Roots: cs.serverTrust1},
+			wantKmStage2: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer2}, Roots: cs.serverTrust2},
+		},
+		{
+			desc: "use identity certs only",
+			options: &PEMFileProviderOptions{
+				CertFile: "not_empty_cert_file",
+				KeyFile:  "not_empty_key_file",
+			},
+			wantKmStage0: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}},
+			wantKmStage1: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}},
+			wantKmStage2: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer2}},
+		},
+		{
+			desc: "use trust certs only",
+			options: &PEMFileProviderOptions{
+				TrustFile: "not_empty_trust_file",
+			},
+			wantKmStage0: certprovider.KeyMaterial{Roots: cs.serverTrust1},
+			wantKmStage1: certprovider.KeyMaterial{Roots: cs.serverTrust1},
+			wantKmStage2: certprovider.KeyMaterial{Roots: cs.serverTrust2},
+		},
+	}
+	for _, test := range tests {
+		test := test
+		testInterval := 200 * time.Millisecond
+		test.options.IdentityInterval = &testInterval
+		test.options.RootInterval = &testInterval
+		t.Run(test.desc, func(t *testing.T) {
+			stage := &stageInfo{}
+			readKeyCertPairFunc = func(certFile, keyFile string) (tls.Certificate, error) {
```

Comment:
Consider following an approach as follows:
```
oldReadKeyCertPairFunc := readKeyCertPairFunc
readKeyCertPairFunc = // define your override here
defer func() {
  readKeyCertPairFunc = oldReadKeyCertPairFunc
}()
```

This will eliminate the need for redefining the read function in `restoreMockFunctionsBack`.
```